truffle/src/main/ruby/core/string.rb

```diff
@@ -1195,72 +1199,55 @@ def match(pattern, pos=0)
     result
   end
 
-  # Removes invalid byte sequences from a String, available since Ruby 2.1.
-  def scrub(replace = nil)
-    output = ''
-    input  = dup
-
-    # The default replacement character is the "Unicode replacement" character.
-    # (U+FFFD).
-    if !replace and !block_given?
-      replace = "\xEF\xBF\xBD".force_encoding("UTF-8")
-                    .encode(self.encoding, :undef => :replace, :replace => '?')
-    end
-
-    if replace
-      unless replace.is_a?(String)
-        raise(
-            TypeError,
-            "no implicit conversion of #{replace.class} into String"
-        )
-      end
-
-      unless replace.valid_encoding?
-        raise(
-            ArgumentError,
-            "replacement must be a valid byte sequence '#{replace.inspect}'"
-        )
-      end
+  def scrub(replace = nil, &block)
+    return dup if valid_encoding?
 
-      replace = replace.force_encoding(Encoding::BINARY)
+    if !replace and !block
+      # The unicode replacement character or '?''
+      replace = "\xEF\xBF\xBD".encode(self.encoding, :undef => :replace, :replace => '?')
     end
 
-    # MRI appears to just return a copy of self when the input encoding is
-    # BINARY/ASCII_8BIT.
-    if input.encoding == Encoding::BINARY
-      return input
+    taint = tainted?
+    replace_block = if replace
+      replace = str_compat_and_valid(replace, encoding)
+      taint = true if replace.tainted?
+      Proc.new { |broken| replace }
+    else
+      Proc.new { |broken|
+        replacement = block.call(broken)
+        replacement = str_compat_and_valid(replacement, encoding)
+        taint = true if replacement.tainted?
+        replacement
+      }
     end
 
-    converter = Encoding::Converter.new(input.encoding, Encoding::BINARY)
-
-    while input.length > 0
-      result = converter.primitive_convert(input, output, output.length)
-
-      if result == :finished
-        break
-      elsif result == :undefined_conversion
-        output << converter.primitive_errinfo[3]
-      else
-        # Blocks can return strings in any encoding so we'll make sure it's the
-        # same as our buffer for the mean time.
-        if block_given?
-          block_output = yield(converter.primitive_errinfo[3])
-
-          output << block_output.force_encoding(output.encoding)
-        else
-          output << replace
-        end
-      end
+    val = Truffle.invoke_primitive(:string_scrub, self, nil, replace_block)
+    if val.nil?
+      dup
+    else
+      val.taint if taint
+      val
     end
-
-    return output.force_encoding(encoding)
   end
 
   def scrub!(replace = nil, &block)
     replace(scrub(replace, &block))
     return self
   end
 
+  def str_compat_and_valid(str, enc)
+    res = StringValue(str)
+    if !res.valid_encoding?
+      raise ArgumentError, "replacement must be valid byte sequence"
+    else
+      if res.ascii_only? ? !enc.ascii_compatible? : enc != res.encoding
```

Comment:
Looking at this more, it's not clear why this isn't just `Encoding.compatible?` or `Rubinius::Type.compatible_encoding`. There's a bit of a logical difference here, since `Encoding.compatible?` checks if strings are empty as well. I can't easily tell if the behavioral difference is intended or is a subtle bug.